clustering/__init__.py

```diff
@@ -4,53 +4,25 @@
 from sklearn.decomposition import PCA
 from sklearn.cluster import KMeans
 
-
-def find_incident(labels):
-    res = set()
-    for i in range(len(labels)):
-        if labels[i].startswith('i_'):
-            res.add(i)
-    return res
-
-
-# numpy array construction
-def construct_np_allin(allin, m, n, incident_labels):
-    np_allin = []
-    for i in range(m):
-        tmp = []
-        count = 0
-        for j in range(n):
-            if allin[i][j] == True:
-                tmp.append(1)
-            elif allin[i][j] == False:
-                tmp.append(0)
-            elif allin[i][j] != 'None':
-                tmp.append(1)
-            else:
-                tmp.append(0)
-            if j in incident_labels and allin[i][j] == 1:
-                count += 1
-        np_allin.append(tmp+[count])
-    return np_allin
-
+def _create_incidents_matrix(dataframe):
+    false_values = [False, 'False', None, 'None', 0]
+    # Map values to binary
+    transformed = np.isin(dataframe.values, false_values, invert=True)
+    # Add counts of incident labels set to 1
+    counts = np.count_nonzero(dataframe.loc[:, dataframe.columns.str.startswith('i_')], axis=1)
+    counts = counts.reshape(counts.shape[0],-1)
+    # Add the counts as a last column for each entry
+    return np.hstack((transformed, counts))
 
 def cluster(pd_rules):
-    # global incident_labels, m, n, pd_allin
-    np_rules = pd_rules.as_matrix()
-    labels = list(pd_rules.columns.values)
-    allin = np_rules[:, 3:]
```

Comment:
Why are the first 3 columns extracted and not used?

The omitted columns are:
- `upload_time`
- `r_telemetry_rules_plugins_sap_sap_tmpfiles_report`
- `r_telemetry_rules_plugins_shift_ocp_service_denial_report`

I understand why is it so for the first one, not the others (since it continues with similar columns like `r_telemetry_rules_plugins_kernel_qlcnic_driver_crash_report`)

Reply:
thats probably an error. Or maybe the format is different on the anonymized data we have in the datahub?

```
❯ parquet-tools schema data/2018-07-19/rule_data/

message spark_schema {
  optional binary account (UTF8);
  optional binary system_id (UTF8);
  optional int96 upload_time;
  optional boolean c_telemetry_rules_plugins_non_kernel_rpm_database_problems_check_affected_version;
  optional boolean c_telemetry_rules_plugins_networking_vlan_interface_gateway_check_affected_version;
  optional boolean c_telemetry_rules_plugins_networking_neighbour_table_overflow_check_affected_version;
```

Reply:
It seems so. The data I have downloaded doesn't have the `account` and `system_id` column.

Reply:
Correction: It does contain these columns, but for some reason they are placed as last ones.

Reply:
Also, if we are transforming all the data to `True` and `False` values, it doesn't make any difference or bias if the account and system_id columns are kept there. All the values in these columns would be `True` no matter what... Removing columns is more expensive operation then just not paying attention to them.